ckanext/s3filestore/commands.py

```diff
@@ -98,31 +98,110 @@ def upload_all(self):
         try:
             for resource_id, file_path in resource_ids_and_paths.iteritems():
                 resource = connection.execute(text('''
-                    SELECT id, url, url_type
+                    SELECT id, url
                     FROM resource
                     WHERE id = :id
+                    AND url IS NOT NULL
+                    AND url <> ''
+                    AND url_type = 'upload'
                 '''), id=resource_id)
                 if resource.rowcount:
-                    _id, url, _type = resource.first()
-                    if _type == 'upload' and url:
-                        file_name = url.split('/')[-1] if '/' in url else url
-                        resource_ids_and_names[_id] = file_name.lower()
+                    _id, url = resource.first()
+                    file_name = url.split('/')[-1] if '/' in url else url
+                    resource_ids_and_names[_id] = file_name.lower()
+                else:
+                    print("{} is an orphan; no resource points to it".format(file_path))
         finally:
             connection.close()
             engine.dispose()
 
         print('{0} resources matched on the database'.format(
             len(resource_ids_and_names.keys())))
 
-        s3_connection = boto3.resource('s3', aws_access_key_id=AWS_ACCESS_KEY_ID, aws_secret_access_key=AWS_SECRET_ACCESS_KEY)
+        _upload_files_to_s3(resource_ids_and_names, resource_ids_and_paths)
+
+    def upload_pairtree(self):
+        def _to_pairtree_path(path):
+            return os.path.join(*[path[i:i + 2] for i in range(0, len(path), 2)])
 
-        uploaded_resources = []
-        for resource_id, file_name in resource_ids_and_names.iteritems():
-            key = 'resources/{resource_id}/{file_name}'.format(
-                resource_id=resource_id, file_name=file_name)
+        BASE_PATH = os.path.join(
+            config.get('ckan.storage_path', config.get('ofs.storage_dir', '/var/lib/ckan/default')),
+            'pairtree_root',
+            _to_pairtree_path(config.get('ckan.storage.key_prefix', 'ckan-file')),
```

Comment:
where did you base did code from?

Reply:
You mean the algorithm to split the string and insert slashes? Found it on Stack Overflow.